minor-modes.lisp

```diff
@@ -0,0 +1,1083 @@
+                                                 rest))))
+                           (bind-it key))))
+                      ((null rest)
+                       (bind-it key))
+                      (t (let ((m (stumpwm:make-sparse-keymap)))
+                           (bind-it key m)
+                           (setf curmap m)))))
+                  (traverse-and-bind (seq)
+                    (loop for (key . rest) on (cl-ppcre:split " " seq)
+                          do (let ((bind (stumpwm:lookup-key curmap
+                                                             (stumpwm:kbd key))))
+                               (attempt-binding key rest bind seq)))))
+               (if (not (or (symbolp bind-to)
+                            (stringp bind-to)
+                            (functionp bind-to)))
+                   (restart-case (error "Invalid binding ~A" bind-to)
+                     (bind-anyway ()
+                       :report "Bind the key binding regardless"
+                       (traverse-and-bind in-seq))
+                     (skip-binding ()
+                       :report "skip this binding"
+                       nil))
+                   (traverse-and-bind in-seq))))))
+      (cond ((null keymap)
+             topmap)
+            ((or (symbolp keymap)
+                 (stumpwm::kmap-p keymap))
+             keymap)
+            ((listp keymap)
+             (restart-case (mapc (lambda (keys)
+                                   (create-keymap-binding keys)
+                                   (setf curmap topmap))
+                                 keymap)
+               (abort-bindings ()
+                 :report "Return the keymap without binding further keys"
+                 topmap)
+               (abort-bindings* ()
+                 :report "Return an empty keymap"
+                 (stumpwm:make-sparse-keymap)))
+             topmap)
+            (t (restart-case
+                   (error "Function MAKE-MINOR-MODE-KEYMAP cant understand ~A"
+                          keymap)
+                 (use-empty-keymap ()
+                   :report "Use an empty keymap"
+                   (stumpwm:make-sparse-keymap))))))))
+
+(defun make-minor-mode-keymap (spec)
+  (generate-keymap spec))
+
+(defun make-minor-mode-top-map (top-map-spec root-map-spec)
+  "Create a top map for a minor mode based upon its TOP-MAP-SPEC and
+ROOT-MAP-SPEC."
+  (let ((top-map nil)
+        (root-map (if root-map-spec
+                      (make-minor-mode-keymap root-map-spec)
+                      (make-sparse-keymap))))
+    (fill-keymap top-map *escape-key* root-map)
+    (generate-keymap top-map-spec top-map)))
+
+
+(eval-when (:compile-toplevel :load-toplevel :execute)
+  (defun make-special-variable-name (mode name)
+    (intern (format nil "*~A-~A*" mode name)))
+  
+  (defun parse-minor-mode-options (options)
+    (let ((valid-options
+            '((:interactive            1)
```

Comment:
```list
            '((:interactive      .     1)
...
```

A alist is a list of conses, not a list of lists. That is, each element is a single cons, with the car being the key and the cdr the value. You are using assoc down below, which is meant to operate on alists, so I think you intended to define an alist

Reply:
While it currently is an alist whose values are lists of length 1, I wasnt sure when writing all of this if more values would be useful for the options (eg type information). Ill change this to a proper alist given that I dont see other information being particularly useful.